programs/clearing_house/src/math/margin.rs

```diff
@@ -1,91 +1,260 @@
 use crate::error::ClearingHouseResult;
 use crate::math::collateral::calculate_updated_collateral;
-use crate::math::constants::{BANK_WEIGHT_PRECISION, MARGIN_PRECISION};
+use crate::math::constants::{
+    BANK_IMF_PRECISION, BANK_WEIGHT_PRECISION, BID_ASK_SPREAD_PRECISION_I128, MARGIN_PRECISION,
+};
 use crate::math::position::{
-    calculate_base_asset_value, calculate_base_asset_value_and_pnl,
-    calculate_base_asset_value_and_pnl_with_oracle_price, calculate_position_pnl,
+    calculate_base_asset_value_and_pnl, calculate_base_asset_value_and_pnl_with_oracle_price,
+    calculate_base_asset_value_with_oracle_price,
 };
 use crate::math_error;
 use crate::state::user::User;
 
 use crate::error::ErrorCode;
 use crate::math::amm::use_oracle_price_for_margin_calculation;
-use crate::math::bank_balance::get_balance_value;
+use crate::math::bank_balance::get_balance_value_and_token_amount;
 use crate::math::casting::cast_to_i128;
 use crate::math::oracle::{get_oracle_status, OracleStatus};
 // use crate::math::repeg;
 use crate::math::slippage::calculate_slippage;
+use crate::state::bank::Bank;
 use crate::state::bank::BankBalanceType;
 use crate::state::bank_map::BankMap;
+use crate::state::market::Market;
 use crate::state::market_map::MarketMap;
+use crate::state::oracle::OraclePriceData;
 use crate::state::oracle_map::OracleMap;
 use crate::state::state::OracleGuardRails;
+use crate::state::user::{MarketPosition, UserBankBalance};
 use crate::validate;
+use num_integer::Roots;
 use solana_program::msg;
+use std::cmp::{max, min};
 use std::ops::Div;
 
-#[derive(Copy, Clone)]
+#[derive(Clone, Copy, PartialEq, Debug)]
 pub enum MarginRequirementType {
     Initial,
     Partial,
     Maintenance,
 }
 
+pub fn calculate_size_markup_liability_weight(
```

Comment:
nit: what do you think about `premium` instead of `markup`? Premium seems more akin to discount, but i realize it's subjective and mark up works too